File: extension/src/channels/routing.ts
import type { RouteSelection, RoutingMode, UUID } from '../shared/types';

export interface CheckboxRule {
  sendDisabled: boolean;
  receiveDisabled: boolean;
  sendTooltip?: string;
  receiveTooltip?: string;
}

export function routingMode(routes: readonly RouteSelection[]): RoutingMode {
  const send = routes.filter((route) => route.sendEnabled);
  const receive = routes.filter((route) => route.receiveEnabled);
  if (send.length === 0 && receive.length === 0) return 'inactive';
  if (send.length > 0 && receive.length === 0) return 'send-only';
  if (send.length === 0 && receive.length > 0) return 'receive-only';
  if (send.length === 1 && receive.length === 1 && send[0]?.channelId === receive[0]?.channelId) return 'sync';
  throw new Error('Invalid ClipMesh routing state');
}

export function isValidRouting(routes: readonly RouteSelection[]): boolean {
  try { routingMode(routes); return true; } catch { return false; }
}

export function checkboxRules(routes: readonly RouteSelection[], channelName:(id:UUID)=>string=(id)=>id): Map<UUID, CheckboxRule> {
  const mode = routingMode(routes);
  const send = routes.filter((route) => route.sendEnabled);
  const receive = routes.filter((route) => route.receiveEnabled);
  return new Map(routes.map((route) => {
    const rule: CheckboxRule = { sendDisabled: false, receiveDisabled: false };
    if (mode === 'send-only') {
      rule.receiveDisabled = !(send.length === 1 && route.channelId === send[0]?.channelId);
      if (rule.receiveDisabled) rule.receiveTooltip = 'To enable Receive, uncheck Send on all channels except this one, or uncheck all Send channels to enter receive-only mode.';
    } else if (mode === 'receive-only') {
      rule.sendDisabled = !(receive.length === 1 && route.channelId === receive[0]?.channelId);
      if (rule.sendDisabled) rule.sendTooltip = 'To enable Send, uncheck Receive on all channels except this one, or uncheck all Receive channels to enter send-only mode.';
    } else if (mode === 'sync') {
      const synchronized = send[0]?.channelId;
      rule.sendDisabled = route.channelId !== synchronized;
      rule.receiveDisabled = route.channelId !== synchronized;
      if (rule.sendDisabled&&synchronized) rule.sendTooltip = `To send to multiple channels, first turn off Receive on ${channelName(synchronized)}.`;
      if (rule.receiveDisabled&&synchronized) rule.receiveTooltip = `To receive from multiple channels, first turn off Send on ${channelName(synchronized)}.`;
    }
    return [route.channelId, rule];
  }));
}

export function toggleRoute(routes: readonly RouteSelection[], channelId: UUID, side: 'send' | 'receive', enabled: boolean): RouteSelection[] {
  const rules = checkboxRules(routes);
  const rule = rules.get(channelId);
  if (!rule) throw new Error('Unknown channel');
  if (enabled && (side === 'send' ? rule.sendDisabled : rule.receiveDisabled)) throw new Error(`${side} checkbox is disabled`);
  const next = routes.map((route) => route.channelId === channelId ? {
    ...route,
    sendEnabled: side === 'send' ? enabled : route.sendEnabled,
    receiveEnabled: side === 'receive' ? enabled : route.receiveEnabled,
  } : route);
  if (!isValidRouting(next)) throw new Error('Transition would produce an invalid routing state');
  return next;
}

export function addJoinedChannel(routes: readonly RouteSelection[], channelId: UUID): RouteSelection[] {
  if (routes.some((route) => route.channelId === channelId)) return [...routes];
  const first = routes.length === 0;
  return [...routes, { channelId, sendEnabled: first, receiveEnabled: first }];
}

export function removeJoinedChannel(routes: readonly RouteSelection[], channelId: UUID): RouteSelection[] {
  const next = routes.filter((route) => route.channelId !== channelId);
  if (!isValidRouting(next)) throw new Error('Removing channel produced invalid routing state');
  return next;
}

File: crates/clipmesh-protocol/src/routing.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RouteSelection {
    pub channel_id: Uuid,
    pub send_enabled: bool,
    pub receive_enabled: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum RoutingMode {
    Inactive,
    SendOnly,
    ReceiveOnly,
    Sync,
}

pub fn routing_mode(routes: &[RouteSelection]) -> Result<RoutingMode, &'static str> {
    let send: Vec<_> = routes.iter().filter(|route| route.send_enabled).collect();
    let receive: Vec<_> = routes
        .iter()
        .filter(|route| route.receive_enabled)
        .collect();
    match (send.len(), receive.len()) {
        (0, 0) => Ok(RoutingMode::Inactive),
        (_, 0) => Ok(RoutingMode::SendOnly),
        (0, _) => Ok(RoutingMode::ReceiveOnly),
        (1, 1) if send[0].channel_id == receive[0].channel_id => Ok(RoutingMode::Sync),
        _ => Err("invalid ClipMesh routing state"),
    }
}

pub fn add_channel(routes: &mut Vec<RouteSelection>, channel_id: Uuid) {
    if routes.iter().any(|route| route.channel_id == channel_id) {
        return;
    }
    let first = routes.is_empty();
    routes.push(RouteSelection {
        channel_id,
        send_enabled: first,
        receive_enabled: first,
    });
}

pub fn set_route(
    routes: &mut [RouteSelection],
    channel_id: Uuid,
    send: Option<bool>,
    receive: Option<bool>,
) -> Result<(), &'static str> {
    let previous = routes.to_vec();
    let route = routes
        .iter_mut()
        .find(|route| route.channel_id == channel_id)
        .ok_or("unknown channel")?;
    if let Some(value) = send {
        route.send_enabled = value;
    }
    if let Some(value) = receive {
        route.receive_enabled = value;
    }
    if routing_mode(routes).is_err() {
        routes.clone_from_slice(&previous);
        return Err("transition would produce an invalid routing state");
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn permits_only_the_four_routing_modes() {
        let a = Uuid::new_v4();
        let b = Uuid::new_v4();
        let mut routes = vec![
            RouteSelection {
                channel_id: a,
                send_enabled: true,
                receive_enabled: true,
            },
            RouteSelection {
                channel_id: b,
                send_enabled: false,
                receive_enabled: false,
            },
        ];
        assert_eq!(routing_mode(&routes).unwrap(), RoutingMode::Sync);
        assert!(set_route(&mut routes, b, Some(true), None).is_err());
        assert_eq!(routing_mode(&routes).unwrap(), RoutingMode::Sync);
    }
}
